omniscidb/QueryEngine/Execute.cpp

```diff
@@ -1796,6 +1813,79 @@ hdk::ResultSetTable Executor::finishStreamExecution(
   return result;
 }
 
+std::pair<std::unique_ptr<policy::ExecutionPolicy>, ExecutorDeviceType>
+Executor::getExecutionPolicyForTargets(const RelAlgExecutionUnit& ra_exe_unit,
+                                       const ExecutorDeviceType requested_device_type,
+                                       const std::vector<InputTableInfo>& query_infos,
+                                       size_t& max_groups_buffer_entry_guess) {
+  if (isDevicesRestricted(ra_exe_unit, requested_device_type)) {
+    LOG(DEBUG1) << "Devices Restricted, falling back on CPU";
+    return {std::make_unique<policy::FragmentIDAssignmentExecutionPolicy>(
+                ExecutorDeviceType::CPU),
+            ExecutorDeviceType::CPU};
+  }
+
+  CHECK(!query_infos.empty());
+  if (!max_groups_buffer_entry_guess) {
+    LOG(DEBUG1) << "The query has failed the first execution attempt because of running "
+                   "out of group by slots. Make the conservative choice: allocate "
+                   "fragment size slots and run on the CPU.";
+    max_groups_buffer_entry_guess = compute_buffer_entry_guess(query_infos);
+    return {std::make_unique<policy::FragmentIDAssignmentExecutionPolicy>(
+                ExecutorDeviceType::CPU),
+            ExecutorDeviceType::CPU};
+  }
+
+  std::unique_ptr<policy::ExecutionPolicy> exe_policy;
+  auto cfg = config_->exec.heterogeneous;
+
+  if (config_->exec.use_cost_model && ra_exe_unit.cost_model != nullptr &&
+      ra_exe_unit.templ != costmodel::AnalyticalTemplate::Unknown &&
+      cfg.enable_heterogeneous_execution) {
+    LOG(DEBUG1) << "Cost Model enabled, making prediction for template "
+                << costmodel::templateToString(ra_exe_unit.templ);
+    size_t bytes = 0;
+
+    // TODO how can we get bytes estimation more correctly?
+    for (const auto& e : query_infos) {
+      auto t = e.info;
+      for (const auto& f : t.fragments) {
+        for (const auto& [k, v] : f.getChunkMetadataMapPhysical()) {
+          bytes += v->numBytes();
+        }
+      }
+    }
+
+    costmodel::QueryInfo qi = {.templ = ra_exe_unit.templ, .bytesSize = bytes};
+
+    // TODO check that template is available for cost model
+    exe_policy = ra_exe_unit.cost_model->predict(qi);
+    return {std::move(exe_policy), requested_device_type};
+  }
+
+  LOG(DEBUG1) << "Cost Model disabled, or template is unknown: templ="
+              << costmodel::templateToString(ra_exe_unit.templ)
```

Comment:
There's a common mechanism for all the to-string methods in HDK, please have a look at https://github.com/intel-ai/hdk/blob/main/omniscidb/Shared/toString.h.